### PR TITLE
Add tutorials

### PR DESCRIPTION
**Pull request by spyridon97**
This PR fixes several documentation warnings, and adds tutorials for the basic ADIOS concepts

This PR fixes https://github.com/ornladios/ADIOS2/issues/3793.

### CODE REVIEW

File: docs/user_guide/source/tutorials/timesteps.rst [hunk 1]
```diff
@@ -0,0 +1,203 @@
+Timesteps
+=========
+
+.. _sec:tutorials_basics_timesteps:
+
+In the previous tutorial, we introduced the concept of operators, and briefly touched upon the concept of timesteps.tutorial
```
Comment (anagainaru) [inline docs/user_guide/source/tutorials/timesteps.rst:6]: Is this a typo (the ".tutorial" at the end of this sentence)?

**Review by anagainaru — COMMENTED** [review]
Two questions/comments
1. You moved a bunch of examples from bpWriteRead{name} to bpTimeWriteRead{name}. What is the significance of Time in the name? I would keep the name the way it was just because it was shorter and still intuitive (unless there is a reason for the Time addition).
2. You added a bunch of skeleton codes (which is great for the tutorial text you have). I would place them all in a separate folder instead of the hello 

I left a bunch of comments with a few typos, but other than this it looks really good. Good job !

**Comment by spyridon97** [comment]
> 1. You moved a bunch of examples from bpWriteRead{name} to bpTimeWriteRead{name}. What is the significance of Time in the name? I would keep the name the way it was just because it was shorter and still intuitive (unless there is a reason for the Time addition).

I did it because bpWriteRead is kinda very vague, and there are a bunch of examples like that. This particular examples show how to write and read time steps. And that's how i also use them.

> 2. You added a bunch of skeleton codes (which is great for the tutorial text you have). I would place them all in a separate folder instead of the hello

@vicentebolea and I decided that the best place to put them is next to the actual code so that we don't forget to update them, if needed.

**Comment by eisenhauer** [comment]
> 
> So you would prefer to rename the examples to bpStepWriteRead*?
Definitely would be my preference.  "Time" doesn't appear in our APIs, but "Step" is prominent.  ("Timestep" is in some SST engine parameters, but we likely should change that at some point.)

> And is it okay if i stick to timesteps for the name of the particular tutorial?
I'd ask @pnorbert to weigh in on that.  I know about the de-emphasis on Timestep in favor of Step, but not as much about the why's.  Norbert has done the demo and tutorials enough to opine on this.

**Comment by spyridon97** [comment]
> > So you would prefer to rename the examples to bpStepWriteRead*?
> > Definitely would be my preference.  "Time" doesn't appear in our APIs, but "Step" is prominent.  ("Timestep" is in some SST engine parameters, but we likely should change that at some point.)
> 
> > And is it okay if i stick to timesteps for the name of the particular tutorial?
> > I'd ask @pnorbert to weigh in on that.  I know about the de-emphasis on Timestep in favor of Step, but not as much about the why's.  Norbert has done the demo and tutorials enough to opine on this.

I decided to remove the word from everywhere in this set of tutorials, I hope addresses all of your concerns helps.

File: docs/user_guide/source/components/engine.rst [hunk 2]
```diff
@@ -323,7 +325,7 @@ PerformsPuts
 
 
 PerformsDataWrite
```
Comment (eisenhauer) [inline docs/user_guide/source/components/engine.rst:327]: Not an issue with this update, but everywhere that says "Performs*" in headers here should be "Perform" (without the s).

Comment (spyridon97) [reply docs/user_guide/source/components/engine.rst]: i fixed that too.

**Comment by vicentebolea** [comment]
The crusher failure can be resolve if you "update the branch", _IMPORTANT_ use rebase update

**Comment by vicentebolea** [comment]
Screenshot from 2023-11-01 12-35-21

**Comment by spyridon97** [comment]
@vicentebolea the bpls results were added, can you check them and approve this PR?

**Review by vicentebolea — COMMENTED** [review]
The changes with regards the bpls usage looks good. :+1: 

As for the the several changes in the documentation unrelated to the tutorials, they look good in a first glance, however, while it is great to improve things here and there, it is best to do it in a different PR. The reason for this is that opening many fronts will both slow the review process and confuse reviewers.

**Comment by spyridon97** [comment]
> The changes with regards the bpls usage looks good. 👍
> 
> As for the the several changes in the documentation unrelated to the tutorials, they look good in a first glance, however, while it is great to improve things here and there, it is best to do it in a different PR. The reason for this is that opening many fronts will both slow the review process and confuse reviewers.

I will keep that in mind for future PRs.

File: docs/user_guide/source/advanced/query.rst [hunk 3]
```diff
@@ -21,14 +21,14 @@ to construct a query and evaluate using the engine.
         // configFile has query, can be either xml or json
         QueryWorker(const std::string &configFile, adios2::Engine &engine);
 
-	// touched_blocks is a list of regions specified by (start, count),
-	// that contains data that satisfies the query file
+	      // touched_blocks is a list of regions specified by (start, count),
+	      // that contains data that satisfies the query file
```
Comment (vicentebolea) [inline docs/user_guide/source/advanced/query.rst:25]:
```suggestion
	    // touched_blocks is a list of regions specified by (start, count),
	    // that contains data that satisfies the query file
```

File: docs/user_guide/source/components/engine.rst [hunk 4]
```diff
@@ -361,13 +363,13 @@ The following table summarizes the memory contracts required by ADIOS2 engines b
 +----------+-------------+-----------------------------------------------+
 | Get      | Data Memory | Contract                                      |
 +----------+-------------+-----------------------------------------------+
-|          | Pointer     | do not modify until PerformPuts/EndStep/Close |
+|          | Pointer     | do not modify until PerformGets/EndStep/Close |
```
Comment (vicentebolea) [inline docs/user_guide/source/components/engine.rst:366]: Is this correct? Shouldnt it be Put?

File: docs/user_guide/source/ecosystem/h5vol/vol.rst [hunk 5]
```diff
@@ -1,11 +1,12 @@
-***********
+**********
 Disclaimer
-***********
+**********
+
 .. note::
 
-The Virtual Object Layer (VOL) is a feature introduced in recent release of  HDF5 1.12 (https://hdf5.wiki/index.php/New_Features_in_HDF5_Release_1.12). 
+  The Virtual Object Layer (VOL) is a feature introduced in recent release of  HDF5 1.12 (https://hdf5.wiki/index.php/New_Features_in_HDF5_Release_1.12).
```
Comment (vicentebolea) [inline docs/user_guide/source/ecosystem/h5vol/vol.rst:7]: Why this space at the beginning?

File: docs/user_guide/source/advanced/ecp_hardware.rst [hunk 6]
```diff
@@ -33,27 +33,29 @@ your tests.
...
- # Launch simple writer test instance
- srun {PROJFLAGS }-N 1 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonWrite SST mpi_dp_test CPCommPattern=Min,MarshalMethod=BP5'
+  # We omit some of the srun (SLURM) arguments which are specific of the project
+  # you are working on. Note that you could avoid calling srun directly by
+  # setting the CMAKE variable `MPIEXEC_EXECUTABLE`.
 
- # On another terminal launch multiple instances of the Reader test
- srun {PROJFLAGS} -N 2 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonRead SST mpi_dp_test
+  # Launch simple writer test instance
+  srun {PROJFLAGS }-N 1 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonWrite SST mpi_dp_test CPCommPattern=Min,MarshalMethod=BP5
```
Comment (vicentebolea) [inline docs/user_guide/source/advanced/ecp_hardware.rst:42]:
```suggestion
  srun {PROJFLAGS } -N 1 /gpfs/alpine/proj-shared/csc331/vbolea/ADIOS2-build/bin/TestCommonWrite SST mpi_dp_test CPCommPattern=Min,MarshalMethod=BP5
```

Not a change in this PR but good to change :)

**Review by vicentebolea — COMMENTED** [review]
Looks good, couple of comments, let the CI pass before making change so we can see the crusher CI result

**Comment by spyridon97** [comment]
@vicentebolea I believe the changes you suggested could be part of documentation improvements PR. and leave this PR as is.

File: docs/user_guide/source/components/engine.rst (hunk 4 above)
Comment (spyridon97) [reply docs/user_guide/source/components/engine.rst:366]: This is correct. There is another table for PerformPut

File: docs/user_guide/source/ecosystem/h5vol/vol.rst (hunk 5 above)
Comment (spyridon97) [reply docs/user_guide/source/ecosystem/h5vol/vol.rst:7]: Fixed

File: docs/user_guide/source/advanced/query.rst (hunk 3 above)
Comment (spyridon97) [reply docs/user_guide/source/advanced/query.rst:25]: Fixed

File: docs/user_guide/source/advanced/ecp_hardware.rst (hunk 6 above)
Comment (spyridon97) [reply docs/user_guide/source/advanced/ecp_hardware.rst:42]: Fixed

File: examples/hello/CMakeLists.txt [hunk 7]
```diff
@@ -3,31 +3,34 @@
...
 if(ADIOS2_HAVE_CUDA OR ADIOS2_HAVE_Kokkos_CUDA)
-  add_subdirectory(bpWriteReadCuda)
+  add_subdirectory(bpStepsWriteReadCuda)
 endif()
-
 find_package(hip QUIET)
 if(ADIOS2_HAVE_Kokkos_HIP OR hip_FOUND)
-  add_subdirectory(bpWriteReadHip)
+  add_subdirectory(bpStepsWriteReadHip)
 endif()
```
Comment (vicentebolea) [inline examples/hello/CMakeLists.txt:26]: > Not to be addressed here, Note for another PR. 

For consistency across the CMake code we should not use find_package outside of detectoptions if within the project(adios2)

**Review by vicentebolea — APPROVED** [review]
Upon CI passing

**Comment by spyridon97** [comment]
> Upon CI passing

There seems to be unrelated tests failing. Shall i merge?

**Comment by vicentebolea** [comment]
Indeed, it is unrelated, ready to merge!

**Pull request merged**